src/WebJobs.Script/Workers/FunctionDataCache/FunctionDataCache.cs

```diff
@@ -0,0 +1,331 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.Azure.WebJobs.Script.Workers.SharedMemoryDataTransfer;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Workers.FunctionDataCache
+{
+    public class FunctionDataCache : IFunctionDataCache
+    {
+        /// <summary>
+        /// Logger.
+        /// </summary>
+        private readonly ILogger _logger;
+
+        private readonly ISharedMemoryManager _sharedMemoryManager;
+
+        /// <summary>
+        /// Lock to be used with any operations that access state of this cache.
+        /// </summary>
+        private readonly object _lock;
+
+        /// <summary>
+        /// Mapping of <see cref="FunctionDataCacheKey"/> to the <see cref="SharedMemoryMetadata"/>
+        /// which indicates where in shared memory the object exists.
+        /// Note: This must be accessed while holding the <see cref="_lock"/> lock.
+        /// </summary>
+        private readonly Dictionary<FunctionDataCacheKey, SharedMemoryMetadata> _localCache;
+
+        private readonly long _maximumCapacityBytes;
+
+        public FunctionDataCache(ISharedMemoryManager sharedMemoryManager, ILoggerFactory loggerFactory, IEnvironment environment)
+        {
+            _logger = loggerFactory.CreateLogger<FunctionDataCache>();
+            _sharedMemoryManager = sharedMemoryManager;
+            _maximumCapacityBytes = GetMaximumCapacityBytes(environment);
+            _lock = new object();
+            _localCache = new Dictionary<FunctionDataCacheKey, SharedMemoryMetadata>();
+            LRUList = new LinkedList<FunctionDataCacheKey>();
+            ActiveReferences = new Dictionary<FunctionDataCacheKey, long>();
+            RemainingCapacityBytes = _maximumCapacityBytes;
+            IsEnabled = true; // TODO USE THE BELOW CHECK
+            //IsEnabled = GetIsEnabled(environment);
+        }
+
+        public bool IsEnabled { get; private set; }
+
+        internal long RemainingCapacityBytes { get; private set; }
+
+        /// <summary>
+        /// Gets the list maintaining the order of use of <see cref="FunctionDataCacheKey"/> in an LRU fashion.
+        /// The last element of the list is the most recently used item.
+        /// The first element of the list is the least recently used item.
+        /// </summary>
+        internal LinkedList<FunctionDataCacheKey> LRUList { get; private set; }
+
+        internal Dictionary<FunctionDataCacheKey, long> ActiveReferences { get; private set; }
+
+        public bool TryPut(FunctionDataCacheKey cacheKey, SharedMemoryMetadata sharedMemoryMeta, bool isIncrementActiveReference, bool isDeleteOnFailure)
```

Comment:
Given that this would be a blocking call as each caller needs a lock, and if cache is full, we evict unbounded until enough space is available, I am wondering if there could be cases where --
1. First several (> 160,000) items cached are super small (Eg - 100 Bytes each) occupying almost 16 MB of the cache limit.
2. Another function runs that tries to cache say 1 file worth 15 MB.
3. In this case, we'd be blocking all executions until we clear about 150,000ish items in the cache.

This may not be a major issue as eviction seems cheap. But wanted to call out. Perhaps, if we see such issues, we could restrict the number of evictions per `TryPut`.

Reply:
That's a fair point - although from the blob access analysis that we did earlier from production data, it seems a bit like an edge case. But yes, I will add logging here and once we roll this out, it would be good to see the number of evictions we have to do on average.